Also consider refunded item quantity while increasing stock.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

When updating an order, we call `wc_maybe_adjust_line_item_product_stock` to see if any item stock needs adjusting. We were not considering if an item has been refunded, which was causing incorrect stock adjustments.
This fix also takes in to account that an order has been refunded partially or fully.

Fixes #24489

### How to test the changes in this Pull Request:

See #24489

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Also consider refunded item when updating order and adjusting stocks.

vedanshujain committed:
Also consider refunded item quantity while increasing stock.

When updating an order, we call `wc_maybe_adjust_line_item_product_stock` to see if any item stock needs adjusting. We were not considering if an item has been refunded, which was causing incorrect stock adjustments.
This fix also takes in to account that an order has been refunded partially or fully.

Fixes #24489

kloon commented on includes/admin/wc-admin-functions.php:
```diff
@@ -214,21 +214,27 @@ function wc_maybe_adjust_line_item_product_stock( $item, $item_quantity = -1 ) {
+	$item->update_meta_data( '_reduced_stock', $item_quantity + $refunded_item_quantity );
 	$item->save();
 
+	if ( 0 === $diff ) {
```
Perhaps better to move this above the update meta call as if there is no difference no stock should be adjusted?

vedanshujain replied:
you still need to adjust `_reduced_stock` in line 231, so we cannot move it above (even if diff is 0).

kloon replied:
Not sure I understand why it needs to adjust the meta value when no stock adjustment was made? You have the following statement which does the actual stock adjustment

```
if ( $diff < 0 ) {
		$new_stock = wc_update_product_stock( $product, $diff * -1, 'increase' );
	} elseif ( $diff > 0 ) {
		$new_stock = wc_update_product_stock( $product, $diff, 'decrease' );
	}
```
If $diff is 0 not adjust is made, yet you are updating the _reduced_stock meta value even if no adjustment was made.

vedanshujain replied:
Ah you are right, I am presuming that we were updating `_reduced_stock` value in order to maintain consistency, but you don't need to when the diff is 0. 

Updating to return early when diff is 0.

vedanshujain committed:
Return early when diff is 0.

Earlier we were also update  `_reduced_stock` meta everytime, but we don't need to when diff is 0 because this means that everything is already consistent.

kloon reviewed (approved):
LGTM now!
